src/assets.rs

```diff
@@ -28,12 +37,39 @@ pub struct SyntaxReferenceInSet<'a> {
     pub syntax_set: &'a SyntaxSet,
 }
 
+/// Stores and allows lookup of minimal [SyntaxSet]s. The [SyntaxSet]s are
+/// stored in serialized form, and are deserialized on-demand. This gives good
+/// startup performance since only the necessary [SyntaxReference]s needs to be
+/// deserialized.
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub(crate) struct MinimalSyntaxes {
+    /// Lookup the index into `serialized_syntax_sets` of a [SyntaxSet] by the
+    /// name of any [SyntaxReference] inside the [SyntaxSet]
+    /// (We will later add `by_extension`, `by_first_line`, etc.)
+    pub(crate) by_name: HashMap<String, usize>,
+
+    /// Serialized [SyntaxSet]s. Wether or not this data is compressed is
+    /// decided by [COMPRESS_SERIALIZED_MINIMAL_SYNTAXES]
+    pub(crate) serialized_syntax_sets: Vec<Vec<u8>>,
+}
+
 // Compress for size of ~700 kB instead of ~4600 kB at the cost of ~30% longer deserialization time
 pub(crate) const COMPRESS_SYNTAXES: bool = true;
 
 // Compress for size of ~20 kB instead of ~200 kB at the cost of ~30% longer deserialization time
 pub(crate) const COMPRESS_THEMES: bool = true;
 
+// Compress for size of ~400 kB instead of ~2100 kB at the cost of ~30% longer deserialization time
+pub(crate) const COMPRESS_SERIALIZED_MINIMAL_SYNTAXES: bool = true;
+
+// Wether or not to compress the serialized form of [MinimalSyntaxes]. Shall
+// always be `false`, because the data in
+// [MinimalSyntaxes.serialized_syntax_sets] has already been compressed
+// (assuming [COMPRESS_SERIALIZED_MINIMAL_SYNTAXES] is `true`). The "outer" data
+// structures like `by_name` are tiny. If we compress, deserialization can't do
+// efficient byte-by-byte copy of `serialized_syntax_sets`.
```

Comment:
Okay. So this constant `COMPRESS_MINIMAL_SYNTAXES` is only here because we need to pass a `compressed` argument to the `asset_from_*` functions, and we don't want it to be just a plain "magic number" `false`. Correct?

If there is some form of restriction that is should be `false` if `COMPRESS_SERIALIZED_MINIMAL_SYNTAXES` is `true`, maybe we should add a [static assertion](https://docs.rs/static_assertions/1.1.0/static_assertions/macro.const_assert.html) for that? Something like
```rust
const_assert!(COMPRESS_MINIMAL_SYNTAXES ^ COMPRESS_SERIALIZED_MINIMAL_SYNTAXES);
```

(Not sure if it's worth adding a dependency just for that)

Reply:
The constants are there to
* Avoid "magic numbers", like you say. But also to:
* Make sure that the value is in sync across serialize and deserialize
* Make it easy for someone that wants to experiment with getting faster startup by sacrificing binary size to do so

They can all be changed independently, and there is no technical limitation that prevents `COMPRESS_SERIALIZED_MINIMAL_SYNTAXES` from being set to `false`. But if it is false, Rust startup will be ~40 msinstead of ~20 ms IIRC, but the binary size will only shrink by a couple of kB.